### PR TITLE
feat: add longtap event

### PR DESCRIPTION
**Pull request by MillerSvt**
Fixes #

### CODE REVIEW

File: tools/prepare-web-types.mjs [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+const pluginsPath = new URL('../dist/ng-event-plugins/', import.meta.url);
+const webTypesPath = new URL('./web-types.json', pluginsPath);
+const webTypes = await readFile(webTypesPath, {
+    encoding: 'utf8',
+});
+
+const newWebTypes = webTypes.replaceAll(/"\.\/src\/([^"]+)\.ts"/g, (_, matched) => `"./${matched}.d.ts"`);
+
+await writeFile(webTypesPath, newWebTypes);
```
Comment (splincode) [inline tools/prepare-web-types.mjs:11]: why do we need that?

File: projects/ng-event-plugins/web-types.json [hunk 2]
```diff
@@ -43,7 +55,7 @@
                         },
                         {
                             "name": "prevent",
-                            "description": "Call .preventDefault() on the event"
+                            "description": "Call `.preventDefault()` on the event"
```
Comment (splincode) [inline projects/ng-event-plugins/web-types.json:58]: make another PR please for cosmetic changes

File: projects/ng-event-plugins/src/types/event-manager-plugin.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import type {EventManager} from '@angular/platform-browser';
+
+// TODO: Type is exposed since Angular 17, remove when updated
+type EventManagerArg = ConstructorParameters<typeof EventManager>[0][0];
```
Comment (splincode) [inline projects/ng-event-plugins/src/types/event-manager-plugin.ts:4]: Duplicate?

https://github.com/taiga-family/ng-event-plugins/blob/main/projects/ng-event-plugins/src/plugins/abstract.plugin.ts#L4

File: projects/ng-event-plugins/src/types/event-manager-plugin.ts [hunk 4]
```diff
@@ -0,0 +1,6 @@
+import type {EventManager} from '@angular/platform-browser';
```
Comment (splincode) [inline projects/ng-event-plugins/src/types/event-manager-plugin.ts:1]: Add another PR for that changes please

File: README.md [hunk 5]
```diff
@@ -88,6 +88,13 @@ onPinchZoom({ scale }: VisualViewport) {
 }
 ```
 
+5. iOS currently doesn't support the `contextmenu` event. Instead, you can use a custom `longtap` event. This event
+   captures the `contextmenu` event on non-iOS devices and simulates similar behavior on iOS devices.
+
+```html
+<div (longtap)="showContextMenu($event.detail.clientX, $event.detail.clientY)">Div with context menu</div>
```
Comment (splincode) [inline README.md:95]: @waterplea should it be as a plugin instead of directive in taiga-ui?

`<div (tuiLongtap)="showContextMenu($event.detail.clientX, $event.detail.clientY)">Div with context menu</div>`

File: tools/prepare-web-types.mjs (hunk 1 above)
Comment (MillerSvt) [reply tools/prepare-web-types.mjs:11]: All paths in web-types.json are broken after the build. This script will fix that.

File: README.md (hunk 5 above)
Comment (waterplea) [reply README.md:95]: This is a workaround for iOS not properly dispatching `contextmenu` event, so I think it's a good idea to handle this as an event, not with directive, and as part of our supercharging events library so that it can be used even without Taiga UI.

File: projects/ng-event-plugins/src/types/longtap.event.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
+interface LongtapEventInit extends EventInit {
+    clientX: number;
+    clientY: number;
+}
+
+export class LongtapEvent extends CustomEvent<{
+    clientX: number;
+    clientY: number;
+}> {
+    constructor(type: string, {clientX, clientY, ...eventInitDict}: LongtapEventInit) {
```
Comment (waterplea) [inline projects/ng-event-plugins/src/types/longtap.event.ts:10]: Should type be hardcoded `longtap`?

File: projects/ng-event-plugins/src/events/longtap.event.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    private dispatchLongtapEvent(
+        element: HTMLElement,
+        clientX: number,
+        clientY: number,
+    ): void {
+        element.dispatchEvent(
+            new LongtapEvent('longtap', {
+                clientX,
+                clientY,
+                bubbles: false,
```
Comment (waterplea) [inline projects/ng-event-plugins/src/events/longtap.event.ts:136]: Native `contextmenu` event bubbles

File: projects/ng-event-plugins/src/events/longtap.event.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    public supports(event: string): boolean {
+        return /^longtap$/i.test(event);
+    }
+
+    private listenContextmenuEvent(element: HTMLElement): Function {
+        return this.manager.addEventListener(
+            element,
+            'contextmenu.prevent.stop',
```
Comment (waterplea) [inline projects/ng-event-plugins/src/events/longtap.event.ts:41]: Oh, I see you are stopping native `contextmenu` as well. Are we sure we want that?

File: projects/ng-event-plugins/src/events/longtap.event.ts [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+
+        return () => {
+            removeLongtapEventPolyfill();
+
+            element.removeEventListener('longtap', handler);
+        };
+    }
+
+    public supports(event: string): boolean {
+        return /^longtap$/i.test(event);
```
Comment (waterplea) [inline projects/ng-event-plugins/src/events/longtap.event.ts:35]: Why not just `===`?

File: projects/ng-event-plugins/src/utils/is-ios.ts [hunk 10]
```diff
@@ -0,0 +1,10 @@
+import {isApple} from './is-apple';
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply projects/ng-event-plugins/src/utils/is-ios.ts:1]: Yes, here is a copy of these utilities for local use

File: projects/ng-event-plugins/src/types/longtap.event.ts (hunk 6 above)
Comment (MillerSvt) [reply projects/ng-event-plugins/src/types/longtap.event.ts:10]: I don't know. This looks inconsistent with other events, such as `MouseEvent`, `TouchEvent`.

File: projects/ng-event-plugins/src/events/longtap.event.ts (hunk 8 above)
Comment (MillerSvt) [reply projects/ng-event-plugins/src/events/longtap.event.ts:41]: I think so, because there is no way to catch this event in the upstream component. When it listening for a longtap event, the plugin will recreate all listeners.

Comment (waterplea) [reply projects/ng-event-plugins/src/events/longtap.event.ts:41]: Ok, makes sense

File: projects/ng-event-plugins/src/types/longtap.event.ts (hunk 6 above)
Comment (waterplea) [reply projects/ng-event-plugins/src/types/longtap.event.ts:10]: Ok, let's keep it consistent. It's not really meant to be created outside plugin anyway.

File: projects/ng-event-plugins/src/events/longtap.event.ts [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    private listenTouchEvents(element: HTMLElement): Function {
+        let longTapTimeout: any = null;
+        let touchStartCoords: {
+            clientX: number;
+            clientY: number;
+        } | null = null;
+
+        function reset(): void {
```
Comment (waterplea) [inline projects/ng-event-plugins/src/events/longtap.event.ts:55]: Should it be an arrow function for consistency? I don't think we ever used nested functions like that.

**Review by waterplea — APPROVED** [review]

File: projects/ng-event-plugins/src/utils/is-ios.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+
+const IOS_REG_EXP = /ipad|iphone|ipod/;
+
+export function isIos(navigator: Navigator): boolean {
```
Comment (splincode) [inline projects/ng-event-plugins/src/utils/is-ios.ts:5]: Please, wait that PR
https://github.com/taiga-family/ng-web-apis/pull/939

File: tools/prepare-web-types.mjs [hunk 13]
```diff
@@ -0,0 +1,11 @@
+import {readFile, writeFile} from 'fs/promises';
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply tools/prepare-web-types.mjs:1]: For what? These changes are related to the longtap event.

[the remark this answers is not in the document]
Comment (splincode) [reply tools/prepare-web-types.mjs:1]: ok

**Comment by waterplea** [comment]
> Can I merge? Or should we wait for something?

Let's merge and publish
